DOCS-1403: Update README.md (#4)
viamrobotics/module-example-cpp: README.md

# C++ example module

> This example module demonstrates the use of the [C++ SDK](https://github.com/viamrobotics/viam-cpp-sdk) to build a minimal Viam module.

Reviewer: ```suggestion
This example module demonstrates the use of the [C++ SDK](https://github.com/viamrobotics/viam-cpp-sdk) to build a basic Viam module.
```

> It is [available for use on the registry](https://app.viam.com/module/viam/module-example-cpp) and shows how to include CI (Continuous Integration) builds for both x86 and ARM architectures.

Reviewer: ```suggestion
It is [available for use on the Viam registry](https://app.viam.com/module/viam/module-example-cpp) and shows how to include CI (Continuous Integration) builds for both x86 and ARM architectures.
```

You can use this as a reference along with the C++ instructions for [coding your own modules](https://docs.viam.com/registry/create/).

## Requirements

Before building this module, verify your operating system's development environment is compatible.
The tested ideal environment for building this module is Debian Bullseye (used in our CI).

You can also use Ubuntu 22.04 for testing, but the resulting C++ builds will not be binary compatible with Bullseye.

To setup for building this module:

> 1. Install the apt packages in [apt-setup.sh](apt-setup.sh).

Reviewer 2: (q) We've got these `apt install` instructions here but below (at lines 29-31) are another set of instructions. Can these be combined with a single simple command, e.g. `sh ./apt-setup.sh`?

Reviewer: ~If there's no reason to keep these separate, then yes it looks like we could just add the two additional packages to `apt-setup.sh`. I've made a PR doing so if this is the correct thing to do: https://github.com/viamrobotics/module-example-cpp/pull/5~

~Then these docs could indeed just suggest `sh ./apt-setup.sh` for the one-stop shop of dep resolution, both for initial set up and for ongoing usage with the GitHub action.~

EDIT: closed linked PR per Abe's comment. The steps are separate because the one is a runtime requirement while the other is a build requirement.

> 2. Install CMake. Ensure your version of CMake matches the [CMAKE_VERSION in this repository's github workflow](https://github.com/viamrobotics/module-example-cpp/blob/af057c9169a730b2c5de5c26f877c0419fb52c19/.github/workflows/build.yml#L7).

Reviewer 2: (q) do you know if this is necessary? The version you have here is higher than [the minimum required in the SDK](https://github.com/viamrobotics/viam-cpp-sdk/blob/f0d75fdf4e29c61350caa4fe065151161e5190a0/CMakeLists.txt#L35) (3.25).

Reviewer: Agreed. Unless this specific module requires a newer version, I would pin the required CMake version at the same version as the C++ SDK requires, which is 3.25 indeed.

@abe-winter any reason for explicitly requiring of a newer version of CMake (`3.27.4`) here, or is it safe to match the C++ SDK's CMake version requirement (`3.25` + )?

Reviewer 3: the min version in the cmakelists in this repo is 3.7 (i.e. much older than cpp sdk)

3.27.4 isn't a requirement, it is the version that gets downloaded in CI

imo it's okay to just say 'recent version'. most people will get it from snap (linux) or brew (mac) + will get the right thing automatically. also the error message if too old is pretty clear.

only concern from me: make sure to emphasize that you need snap on ubuntu, not apt-get. the apt-get one *is* too old I think

	- To install on Ubuntu 22.04, run `snap install cmake` or follow [CMake's download instructions](https://cmake.org/download/).
3. [Install rust](https://www.rust-lang.org/tools/install) (needed for `viam_rust_utils`).

> ## Build and Run

Reviewer: If we are removing the explicit `sudo apt install` steps from line 30 in favor of merging into the `setup.sh` script per https://github.com/viamrobotics/module-example-cpp/pull/5 , then I'd suggest merging "prepare to build" steps and "build and run" steps into just one section, ending with "then follow the C++ instructions to ..." (line 33).

> ## Build and Run

Reviewer 3: 'build and run' here is probably copying my old instructions, which were low-key wrong

running from registry + running from locally built module are different cases (first one is easy, second one is hard)

imo we should guide people through these two steps for example code (install from reg to see it work, then build), but 1) this is just an opinion, and 2) would require a bit of testing to make sure it's actually doable. (local exec will likely collide with installed one)

Author: Addressing this in the next commit.

Author: Per offline conversation with both you and @stuqdog I updated the Build Locally section to reference complex C++ example module for guidance on how to modify a simple C++ module. Happy to make further changes

Update: After reviews I added a link to the example module thanks to feedback from ethan and @andf-viam

## Configure your Example C++ Module

> [!NOTE]
> Before configuring your example module, you must [create a robot](https://docs.viam.com/manage/fleet/robots/#add-a-new-robot).

Navigate to the **Config** tab of your robot’s page in [the Viam app](https://app.viam.com/).
Click on the **Components** subtab and click **Create component**.
Select **Sensor**, then search for the `wifi:cpp` model. Give your resource a name of your choice, click **Create** and then click **Save config**.

Next, change the `Mode:` selector from `Builder` to `Raw JSON`, and verify that `"components"` array now includes the `viam:wifi:cpp` model and that `"modules"` array includes `viam:module-example-cpp` module. 

Your `Raw JSON` configuration should now include both: 

```json {class="line-numbers linkable-line-numbers"}
{
  "components": [{
    "attributes": {},
		"depends_on": [],
    "name": "wifi-cpp",
		"model": "viam:wifi:cpp",
    "type": "sensor",
		"namespace": "rdk"
  }],
  "modules": [{
		"type": "registry",
	  "name": "viam_module-example-cpp",
    "module_id": "viam:module-example-cpp",

>     "version": "0.0.3"

Reviewer 2: the indenting here looks a little funky, but I don't work with json much! Is this intentional?

Reviewer: Agreed, let's try to match the indentation in the App; this is what I get:

```
{
  "modules": [
    {
      "type": "registry",
      "name": "viam_module-example-cpp",
      "module_id": "viam:module-example-cpp",
      "version": "0.0.3"
    }
  ],
  "components": [
    {
      "name": "my-wifi-cpp",
      "model": "viam:wifi:cpp",
      "type": "sensor",
      "namespace": "rdk",
      "attributes": {},
      "depends_on": []
    }
  ]
}
```

Reviewer 3: also in case this is in here because it was in my old instructions -- we don't need to edit json for registry modules anymore, the UX takes care of it

  }]
}
```

> If you upload your version to the registry after building it, remember to update `model` and `module_id` with your values.

Reviewer: ```suggestion
If you upload your version to the Viam registry after building it, remember to update `model` and `module_id` with your values.
```

> [!NOTE]  
> For more information, see [Configure a Robot](https://docs.viam.com/manage/configuration/).

## Module Contents

You can use this repository's files as a reference when building your C++ module:

| Name  | Content | Location |
| ------------- | ------------- | ------------- |
| `CMakeLists.txt` | The configuration for compiling the binary and managing dependencies. | `-` |
| `Makefile` | The command to create the module tarball dependencies. | `-` |
| `.github/workflows/` | The CI build and deployment logic for the registry. | `-` |
| `run.sh` and `apt-setup.sh` | The entrypoint and dependency installation for running the module on a robot. | `-` |
| `meta.json` | The Viam module configuration. | `-` |
| `main.cpp`  | The module's entry point file.</br>See [Code a main entry point program](https://docs.viam.com/registry/create/#code-a-main-entry-point-program) for more information. | `src`|
| `wifi.cpp` | The resource model which implements all the methods that the Viam RDK require.</br>See [Code a new resource model](https://docs.viam.com/registry/create/#code-a-new-resource-model) for more information. | `src` |

> | `wifi.hpp` | The defined implementation of the derivative `WifiSensor` class in `wifi.cpp`.</br>See [Find your reference files](https://docs.viam.com/registry/create/#find-your-reference-files) for more information. | `src` |

Reviewer 2: this is awesome! 🎉

## Setting Up CI

This module uses [Buildjet runners](https://buildjet.com/for-github-actions) to build natively for ARM. If you fork this repository and don't have Buildjet connected to GitHub, your CI jobs won't start.

You can choose one option for CI setup:
- Exlusively targetting x86
- Building on an ARM emulator using [QEMU](https://www.qemu.org/download/) with [docker/build-push-action](https://github.com/docker/build-push-action)
- Setting up cross-compilation

Refer to the [fork instructions](https://github.com/viam-labs/python-example-module#forking-this-repo) in the Python example module for more details on deploying your fork. 

## Troubleshooting
> [!NOTE]

> > Ensure you have `viam-server` installed on the machine you are using as your robot.

Reviewer: ```suggestion
> Ensure you have [installed `viam-server`](https://docs.viam.com/get-started/installation/) on your machine.
```

### Library Paths

Use `ldd` to identify library path issues.
For example:

> - If you are unable to find some libviam (e.g., libviamsdk.so.noabi), ensure libvimapi + libviamsdk are bundled in your `module.tar.gz`.

Reviewer: ```suggestion
- If you are unable to find one of the `libviam` packages (for example: `libviamsdk.so.noabi`), ensure `libvimapi` and `libviamsdk` are bundled in your `module.tar.gz`.
```

- Confirm that `LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$PWD` (handled by the `run.sh` wrapper) is set.
- If any other libraries are missing, use `apt-get install` to install them.

### Connecting a Debugger

For unexplained issues, you can use a debugger to investigate crashes or disconnects.
The debugger referenced in this case is [GDB](https://www.onlinegdb.com/).

1. Configure your robot to use the module and ensure it's running.
2. Find the PID of your running module with `ps aux | grep module-example-cpp`.

> 3. Run `sudo gdb --pid $PID` with the obtained PID (use `sudo` as the robot may be running in sudo).

Reviewer: ```suggestion
3. Run `sudo gdb --pid $PID` with the obtained PID (use `sudo` as `viam-server` is likely running in sudo).
```

> 4. Type `continue` in your debugggdb to restart the process.

Reviewer: ```suggestion
4. Type `continue` in `gdb` to restart the process.
```

## Build and Run

> You can build this exampl

Reviewer 2: Incomplete sentence here?

These instructions assume a target system similar to Debian Bullseye.

Before you can build this module's executable, you must follow the steps to [prepare a module for execution](https://docs.viam.com/registry/create/#prepare-the-module-for-execution).

> You can also view the [complex C++ module examples](https://github.com/viamrobotics/viam-cpp-sdk/tree/main/src/viam/examples/modules/complex) to see how a simple module like this one can be modified. 

Reviewer 2: (minor) I would update this language slightly. I think "to see how a simple module like this one can be modified" implies that the complex module is structured as a "let's start with something simple and then iterate on it step by step" which isn't quite right. It is true that the complex module does show ways to modify/add to a module! Perhaps you could modify to say, "to see an example of how one could structure and create APIs for a more extensible module."? Or something to that effect.

Reviewer: Great point. I might suggest removing this sentence entirely, actually: this readme is wholly this-module-specific, the broader conceptual intro to modules and their levels of complexity is more the responsibility of the main docs pages under `/registry/`.

Or, if desired, perhaps mention could live at the very end, maybe phrased similarly to:

> For a more involved C++ module example, which defines both a component and service API, see the [C++ complex module example](https://github.com/viamrobotics/viam-cpp-sdk/tree/main/src/viam/examples/modules/complex).

Author: Agreed, I think adding this may thank you both for your feedback! Brief sentence added at the bottom.
